File: Dynamic-Programming/91. Decode Ways/decodeWays.js
var numDecodings = function (s) {
    let n = s.length;
    if (!s || n === 0) return 0;

    let dp = new Array(n + 1).fill(0);
    dp[0] = 1;
    dp[1] = s.charAt(0) === "0" ? 0 : 1;

    for (let i = 2; i < dp.length; i++) {
        let option1 = parseInt(s.substring(i - 1, i));
        let option2 = parseInt(s.substring(i - 2, i));

        if (option1 >= 1 && option1 <= 9) {
            dp[i] += dp[i - 1];
        }
        if (option2 >= 10 && option2 <= 26) {
            dp[i] += dp[i - 2];
        }
    }
    return dp[n];
};

File: Dynamic-Programming/91. Decode Ways/decodeWays.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = [0 for _ in range(len(s) + 1)]
        dp[0] = 1
        dp[1] = 0 if s[0] == "0" else 1
        
        for i in range(2, len(dp)):
            option1 = s[i-1:i]
            option2 = s[i-2:i]
            if 1 <= int(option1) <= 9:
                dp[i] += dp[i - 1]
            if 10 <= int(option2) <= 26:
                dp[i] += dp[i - 2]
            
        return dp[-1]

class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == "0":
            return 0
        
        dp = [0 for _ in range(len(s) + 1)]
        dp[0] = 1
        dp[1] = 0 if s[0] == "0" else dp[0]
        
        for i in range(2, len(dp)):
            option1 = s[i - 1: i]
            option2 = s[i - 2: i]
            if 1 <= int(option1) <= 9:
                dp[i] += dp[i - 1]
            if 10 <= int(option2) <= 26:
                dp[i] += dp[i - 2]
        
        print(dp)
        return dp[-1]